sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -132,7 +132,9 @@ abstract class Optimizer(sessionCatalog: SessionCatalog)
       OptimizeCodegen) ::
     Batch("RewriteSubquery", Once,
       RewritePredicateSubquery,
-      CollapseProject) :: Nil
+      CollapseProject) ::
+    Batch("UpdateAttributeReferences", Once,
```

Comment:
Can we apply this optimization earlier to encourage other optimizations that uses information on `nullable`?
For example, before `Operator Optimizations`?

Or, is there any reason to put this optimization at the end of the optimizations?